pkg/innerring/internal/blockchain/blockchain.go

```diff
@@ -0,0 +1,398 @@
+	cfgBaseProto.Hardforks = opts.HardForks
+
+	cfgBaseApp := &cfgBase.ApplicationConfiguration
+	cfgBaseApp.Relay = true
+	cfgBaseApp.UnlockWallet = cfg.Wallet
+	cfgBaseApp.Consensus.Enabled = true
+	cfgBaseApp.Consensus.UnlockWallet = cfg.Wallet
+	cfgBaseApp.P2PNotary.Enabled = !opts.NotaryDisabled
+	cfgBaseApp.P2PNotary.UnlockWallet = cfg.Wallet
+	cfgBaseApp.StateRoot.UnlockWallet = cfg.Wallet
+	cfgBaseApp.RPC.Enabled = true
+	cfgBaseApp.RPC.Addresses = opts.RPCListenAddresses
+	cfgBaseApp.RPC.StartWhenSynchronized = true
+	cfgBaseApp.RPC.MaxGasInvoke = fixedn.Fixed8FromInt64(100)
+	cfgBaseApp.P2P.Addresses = opts.P2PListenAddresses
+	cfgBaseApp.P2P.DialTimeout = cfg.P2P.DialTimeout
+	cfgBaseApp.P2P.ProtoTickInterval = cfg.P2P.ProtoTickInterval
+	cfgBaseApp.P2P.PingInterval = cfg.P2P.Ping.Interval
+	cfgBaseApp.P2P.PingTimeout = cfg.P2P.Ping.Timeout
+	cfgBaseApp.P2P.MinPeers = int(cfg.P2P.MinPeers)
+	cfgBaseApp.P2P.AttemptConnPeers = int(cfg.P2P.AttemptConnPeers)
+	cfgBaseApp.P2P.MaxPeers = int(cfg.P2P.MaxPeers)
+	// TODO: profiler, metrics
+
+	cfgServer, err := network.NewServerConfig(cfgBase)
+	if err != nil {
+		return nil, fmt.Errorf("compose NeoGo server config from the base one: %w", err)
+	}
+
+	var cfgDB dbconfig.DBConfiguration
+	cfgDB.Type = cfg.Storage.typ
+	if cfgDB.Type == dbconfig.BoltDB {
+		cfgDB.BoltDBOptions.FilePath = cfg.Storage.path
+	} else if cfgDB.Type == dbconfig.LevelDB {
+		cfgDB.LevelDBOptions.DataDirectoryPath = cfg.Storage.path
+	}
+
+	blockchainStorage, err := storage.NewStore(cfgDB)
+	if err != nil {
+		return nil, fmt.Errorf("init storage for blockchain: %w", err)
+	}
+
+	blockChain, err := core.NewBlockchain(blockchainStorage, cfgBase.Blockchain(), opts.Logger)
+	if err != nil {
+		closeErr := blockchainStorage.Close()
+		if closeErr != nil {
+			return nil, fmt.Errorf("init core blockchain component: %w; failed to close blockchain storage: %v", err, closeErr)
+		}
+
+		return nil, fmt.Errorf("init core blockchain component: %w", err)
+	}
+
+	go blockChain.Run()
+
+	netServer, err := network.NewServer(cfgServer, blockChain, blockChain.GetStateSyncModule(), opts.Logger)
+	if err != nil {
+		return nil, fmt.Errorf("init NeoGo network server: %w", err)
+	}
+
+	if !opts.NotaryDisabled {
+		var cfgNotary notary.Config
+		cfgNotary.MainCfg.Enabled = true
+		cfgNotary.MainCfg.UnlockWallet = cfg.Wallet
+		cfgNotary.Chain = blockChain
+		cfgNotary.Log = opts.Logger
+
+		notaryService, err := notary.NewNotary(cfgNotary, netServer.Net, netServer.GetNotaryPool(), func(tx *transaction.Transaction) error {
+			err := netServer.RelayTxn(tx)
+			if err != nil && !errors.Is(err, core.ErrAlreadyExists) {
+				return fmt.Errorf("relay completed notary transaction %s: %w", tx.Hash().StringLE(), err)
+			}
+
+			return nil
+		})
+		if err != nil {
+			return nil, fmt.Errorf("init Notary service: %w", err)
+		}
+
+		netServer.AddService(notaryService)
+		blockChain.SetNotary(notaryService)
```

Comment:
No way, read the comment to `SetNotary`, please:
```
// SetNotary sets notary module. It doesn't protected by mutex and
// must be called before `bc.Run()` to avoid data race.
```
The same thing is for stateroot/oracle/other services.

Reply:
moved `Run` to `Start`